packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -1402,17 +1416,7 @@ class OffsetLayer extends ContainerLayer {
   ///  * [RenderRepaintBoundary.toImage] for a similar API at the render object level.
   ///  * [dart:ui.Scene.toImage] for more information about the image returned.
   Future<ui.Image> toImage(Rect bounds, { double pixelRatio = 1.0 }) async {
-    assert(bounds != null);
-    assert(pixelRatio != null);
-    final ui.SceneBuilder builder = ui.SceneBuilder();
-    final Matrix4 transform = Matrix4.translationValues(
-      (-bounds.left  - offset.dx) * pixelRatio,
-      (-bounds.top - offset.dy) * pixelRatio,
-      0.0,
-    );
-    transform.scale(pixelRatio, pixelRatio);
-    builder.pushTransform(transform.storage);
-    final ui.Scene scene = buildScene(builder);
+    final ui.Scene scene = _createSceneForImage(bounds, pixelRatio: pixelRatio);
```

Comment:
The bounds calculation below is bad math, you don't want the ceil of the size, you want the size of the roundOut()

Reply:
Actually, given that the scene was created by translating back by the origin of the bounds, it might work if there was no offset.

Unfortunately, there is an offset, so the scene isn't necessarily rendered at the origin of the bounds.

So, what is really needed is (bounds+offset).roundOut.size. If offset is zero then this will end up being equivalent to the calculation below, but if it is not zero, then this will fix a potential "clipping off the last pixel" bug.

Reply:
Hmmm, thinking through this more - if the first thing that buildScene does is to translate by offset, then those would cancel out and size.ceil is fine. I don't see that as happening in buildScene, though. Am I missing something there?

Reply:
I need to do some research on the intention of this API and what the engine is actually doing with it.  It seems like it would be reasonable for a developer to specify a fractional offset if they intended to sample from only part of the picture. Which makes ceil not make sense either really. We'd only need to round up to an integral value for the texture/width height.

Reply:
oh wait, its the bounds width / height - isn't that correct then:

developer specifies a Rect.fromLTWH(0.5, 0.5, 1.5, 1.5) -> we create a texture size of (1, 1). developer specifies a Rect.fromLTWH(0.5, 0.5, 1, 1) -> we need to round up to create a texture of (1, 1)?